README.txt
= rubicante

* RubyForge http://rubicante.rubyforge.org
* GitHub http://github.com/avanderhook/rubicante

== DESCRIPTION:

Rubicante is a business natural language (BNL) for System Administrators (SAs).
It allows SAs to define their networks in terms of servers, what those servers
provide (websites, Win32 services), and what ports they listen on.  Once the
network is defined, SAs can type in natural language questions about the network
and receive answers.

== FEATURES/PROBLEMS:

* Describe a network of computer systems
  * Hosts
    * Ports it should be listening on
    * Critical web sites hosted
* Query the network
  * What is wrong: returns items (hosts, websites) that are not functioning

== SYNOPSIS:

There are several ways that you can run Rubicante:
* An interactive shell
* An interactive shell with source file
* Cron mode
* All of the above in DEBUG mode

=== Interactive Shell:

Interactive shell mode allows you to specify hosts and run Rubicante queries against them.

  $ rubicante
  rubicante> Host webserver provides website www.my-intranet.com
  rubicante> Host webserver provides website www.good-website.com, website www.bad-website.com
  rubicante> What is wrong?
  [webserver] is returning code 500 for website www.bad-website.com

=== Interactive Shell with Source File

Interactive shell mode with files allows you to define your network's resources
in a file, load them in to rubicante, and obtain the interactive shell prompt
so you can use Rubicante queries to examine the network.

  $ cat 'my-network.bnl'
  Host webserver listens on port 80, provides website www.example.com
  Host shellserver listens on port 22
  $ rubicante -r my-network.bnl
  rubicante> What is wrong?
  [shellserver] port 22 is closed.

=== Cron Mode

Cron mode reads in the specified file and then asks Rubicante
"what is wrong?".  Any problems discovered will be logged to STDOUT
(making this ideal for calling periodically with cron), and then exiting.

  $ cat 'my-network.bnl'
  host web1 website www.mycompany.com
  host web2 website intranet.mycompany.com
  host web3 website wiki.mycompnay.com
  $ rubicante -c my-network.bnl
  [web2] is returning code 500 for website intranet.mycompany.com
  $

=== DEBUG mode

Use the '-d' flag when executing Rubicante to obtain debug output.

== BASIC LANGUAGE SPECIFICATION:

=== Conventions used
* <parameter> is a required parameter
* [optional] is an optional word (i.e., a "bubble" word for readability)

=== HOST language

* Host <hostname> [provides] website <fqdn>
  * Defines host <hostname> (if not previously defined) and registers the website a <fqdn> with it
  * Example:
  "host www provides website www.example.com
* Host <hostname> [listens] [on] port <port #>
  * Defined host <hostname> (if not previously defined) and registers the port <port #> to it
  * Example:
  "host www listens on port 80"
* COMBINE HOST SUBCOMMANDS IN ONE LINE
  * Example:
  "host www listens on port 80, port 443, provides website www.mycompany.com, website intranet.mycompany.com

=== WHAT language

* What [is] wrong[?]
  * checks all defined hosts for problems and reports them

== REQUIREMENTS:

* Ruby 1.8+
* logging (>= 0.9.7)
* trollop (>= 1.13)

== INSTALL:

* sudo gem install rubicante

== LICENSE:

Copyright (c) 2009, Adam VanderHook
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
* Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
* Neither the name of the Rubicante nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


History.txt
=== 0.0.8 / 2009-10-2

* Updated hash commenting so that a space is no longer necessary (i.e., '#comment' vs. '# comment' now works)
* Support empty lines in files and on the console
* Gracefully handle missing files when the -r flag is used and log an error
* Gracefully handle SocketErrors (typically unresolvable domain names) in website response code checking and log an error
* Better parsing in website URLs
  * Handle http://
  * Handle https:// (with logged warning - we do not support HTTPS requests yet)
  * Handle extened paths
    * Example: host www provides website my.railsapp.com/follow/my/path
    * Will perform an HTTP GET on the path when checking the return code
* Logging improvements
  * Set a new default logging pattern
  * Fixed issue with a rouge newline character in Rubicante::Environment's debug logging

=== 0.0.7 / 2009-9-23

* Added support for hashed comments

=== 0.0.6 / 2009-5-29

* Works with the latest version of the Logging gem

=== 0.0.5 / 2009-04-17

* Fixed a bug where Windows services that were registered on a host but were not running were not reported as a problem

=== 0.0.4 / 2009-04-10

* More gem manifest fixes

=== 0.0.3 / 2009-04-10

* Fix broken gem
  * forgot to include new lib/os_functions.rb

=== 0.0.2 / 2009-04-10

* Support checking on Windows services
  * requires rubicante be run on a Windows host

=== 0.0.1 / 2009-04-08

* Initial alpha release
* Support adding hosts with
  * ports
  * websites
* Report 'what is wrong'


lib/rubicante/host.rb
require 'rubicante/website'
require 'rubicante/host_error'

require 'logging'
require 'ping'
require 'socket'

require 'os_functions'
include OsFunctions
require 'win32ole' if is_windows?

module Rubicante
  class Host
    include OsFunctions

    attr_reader :name
    attr_accessor :ports, :services, :types, :websites

    def initialize(name)
      @name     = name
      @ports    = []
      @services = []
      @types    = []
      @websites = []

      @log = Logging.logger[self]

      # Prepare Website logger
      @appender = Logging.logger['rubicante']
      Logging.logger['Rubicante::Website'].add_appenders(Logging.appenders.stdout) if not @appender.nil?
      Logging.logger['Rubicante::Website'].level = @log.level
    end

    def ping
      @log.debug "Performing TCP echo ping on host '#{@name}'"
      Ping.pingecho @name
    end

    def port(port_number)
      @ports << port_number
      self
    end

    def service(service_name)
      @services << service_name
      self
    end

    def type(type_name)
      @types << type_name
      self
    end

    def website(website_url)
      @websites << Website.new(website_url)
      self
    end

    # Check if the specified port is active by connecting to it
    def check_port(port)
      port_output = "#{@name}:#{port}"

      @log.debug "Checking port #{port_output}..."

      begin
        test = TCPSocket.open(@name, port)
        @log.debug "Port #{port_output} looks good"
        return true # if we get here, the socket opened
      rescue
        @log.debug "Port #{port_output} raised an exception when opening"
        return false  # if we get here, there are problems with the port
      end
    end

    # Iterates through all the ports in the Host and runs check_port(port)
    # against them.
    #
    # == Yields: port, is_alive
    #
    # Each registered port is yielded along with the boolean result from
    # check_port(port).
    #
    # == Example:
    #
    #   host = Rubicante::Host.new("test-host")
    #   host.port(80)
    #   host.port(443)
    #
    #   host.check_ports do |port, response|
    #     puts "Port #{port} is UP" if response
    #     puts "Port #{port} is DOWN" if not response
    #   end
    def check_ports
      @ports.each do |port|
        yield port, check_port(port)
      end
    end

    # Iterates through all the websites in the Host and runs
    # wrong? against them looking for problems.  If a problem is
    # found, a Hash of the URL and the HTTP Status code is yielded.
    #
    # == Yields: a Hash
    #
    #   {
    #     :url    => String,  # the URL of the current website in the block
    #     :code   => String   # the HTTP Status code of the check (i.e., 404, 500, etc.)
    #   }
    #
    # == Example
    #
    #   host = Rubicante::Host.new("test-host")
    #   host.website('www.exmaple.com')
    #   host.website('www.rubicante.com')
    #   host.website('www.openbsd.org')
    #
    #   host.check_websites do |result|
    #     puts "Website #{result[:url]} failed with code #{result[:code]}!"
    #   end
    def check_websites
      @log.debug "Checking websites registered to host '#{@name}'"
      @websites.each do |website|
        yield website.wrong?
      end
    end

    def wrong?
      @log.debug "Determing what is wrong with host '#{@name}'"
      result = HostError.new(@name)
      result.ping = self.ping

      # If the host is alive, continue testing
      if result.ping
        check_ports do |port, response|
          result.bad_ports << port if not response
        end

        check_websites do |website_error|
          result.add(website_error)
        end

        if is_windows?
          check_services do |service, response|
            result.bad_services << service if not response
          end
        end
      end

      return result
    end

    ###
    # Windows-specific functions
    ###
    if is_windows?
      def get_wmi
        WIN32OLE.connect("winmgmts://#{@name}")
      end

      # Checks the host to see if the specified service is running
      def is_running?(service)
        query = "SELECT Name, State FROM Win32_Service WHERE Name='#{service}' AND State='Running'"

        result = false
        service_label = "#{@name}->#{service}"

        @log.debug "Checking services #{service_label}..."
        get_wmi.ExecQuery(query).each do |result|
          @log.debug "#{service_label}.State == #{result.State}"
          result = true if result.State = 'Running'
        end

        return result
      end

      def check_services
        @services.each do |service|
          yield service, is_running?(service)
        end
      end
    end
  end
end
